### ISSUE TITLE
clang: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]

### ISSUE THREAD

**Issue by anatoli26** [cyrusimap/cyrus-imapd#3265]
When building with clang, there are 37 warnings during the build process about unknown attribute 'optimize':
```
imap/index.c:6408:41: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
    __attribute__((pure, always_inline, optimize("-O3")));
                                        ^
```

Is there a way to make it compatible with clang? On the other hand, should these optimizations be there? For example, OpenBSD doesn't support optimizations higher than O2.

**Comment by anatoli26** [comment]
@elliefm, any idea how to solve this? After completing the format specifier fixes, this (the `'optimize' ignored`) warning is the last massive one on OpenBSD 

There are a couple of other warnings repeated 4 and 1 times respectively:
```
Makefile:374: warning: ignoring prerequisites on suffix rule definition
Makefile:374: warning: ignoring prerequisites on suffix rule definition
In file included from IMAP.xs:56:
In file included from ../../lib/util.h:49:
../../config.h:618:9: warning: 'VERSION' macro redefined [-Wmacro-redefined]
#define VERSION "3.2.4"
        ^
<command line>:3:9: note: previous definition is here
#define VERSION "1.00"
        ^
1 warning generated.
Makefile:363: warning: ignoring prerequisites on suffix rule definition
Makefile:363: warning: ignoring prerequisites on suffix rule definition
```

**Comment by elliefm** [comment]
> `imap/index.c:6408:41: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
    __attribute__((pure, always_inline, optimize("-O3")));`
> Is there a way to make it compatible with clang? On the other hand, should these optimizations be there? For example, OpenBSD doesn't support optimizations higher than O2.

I don't know much about clang.  It's not currently clear to me whether this warning is telling us that clang does not know about `__attribute__((optimize))` at all, or whether it's telling us that the target platform (OpenBSD) does not support optimisation levels greater than 2...

We already have a test for this in configure.ac and all the optimize attributes are (or should be) gated behind it.  After you run configure, what does `grep HAVE_DECLARE_OPTIMIZE config.h` find?

> `Makefile:374: warning: ignoring prerequisites on suffix rule definition`

I'm not sure what this means, but Makefile is generated by automake so it's somewhat out of our control.  Are you using the BSD `make`?  Does using `gmake` instead fix this?  What does line 374 of the generated Makefile contain?

```
In file included from IMAP.xs:56:
In file included from ../../lib/util.h:49:
../../config.h:618:9: warning: 'VERSION' macro redefined [-Wmacro-redefined]
#define VERSION "3.2.4"
        ^
<command line>:3:9: note: previous definition is here
#define VERSION "1.00"
        ^
1 warning generated.
```

Nothing we can do about this one.  Autotools, and the perl ExtUtils::MakeMaker build system for XS modules (i.e. perl modules written in C), both want to define VERSION for their own purposes.  The Cyrus::IMAP perl module links against libcyrus, and therefore sees both during compilation.  It is at least using the correct one (3.2.4), overriding the junk one (1.0.0). We've been ignoring this warning for years, and you should too. :)

**Comment by anatoli26** [comment]
> I don't know much about clang. It's not currently clear to me whether this warning is telling us that clang does not know about `__attribute__((optimize))` at all, or whether it's telling us that the target platform (OpenBSD) does not support optimisation levels greater than 2...

It's saying it doesn't recognize this attribute at all.

> We already have a test for this in configure.ac and all the optimize attributes are (or should be) gated behind it. After you run configure, what does `grep HAVE_DECLARE_OPTIMIZE config.h` find?
```
#define HAVE_DECLARE_OPTIMIZE /**/
```

> Are you using the BSD `make`?

Yes

>  What does line 374 of the generated Makefile contain?

```
 371 #am__append_76 = $(LD_SIEVE_ADD)
 372 ##am__append_77 = cunit/jmap_util.testc
 373 am__append_78 = \
 374     imap/autocreate.c \
 375     imap/autocreate.h
 376
 377 #am__append_79 = lib/cyrusdb_sql.c
 378 am__append_80 = lib/sqldb.c
 379 am__append_81 = lib/gmtoff_tm.c
 380 #am__append_82 = lib/gmtoff_gmtime.c
 381 am__append_83 = lib/nonblock_fcntl.c
 382 #am__append_84 = lib/nonblock_ioctl.c
 383 #am__append_85 = lib/cyrusdb_zeroskip.c
 384 #am__append_86 = $(ZEROSKIP_LIBS)
 385 #am__append_87 = $(ZEROSKIP_CFLAGS)
 386 #am__append_88 = $(ZEROSKIP_CFLAGS)
 387 #am__append_89 = lib/xsha1.c
 388 #am__append_90 = lib/getaddrinfo.c
 389 #am__append_91 = lib/getnameinfo.c
 390 am__append_92 = lib/lock_fcntl.c
 391 #am__append_93 = lib/lock_flock.c
 392 #am__append_94 = lib/map_shared.c
 393 #am__append_95 = lib/map_stupidshared.c
 394 am__append_96 = lib/map_nommap.c
 395 #am__append_97 = \
 396 #   man/squatter.8
 397
 398 #am__append_98 = \
 399 #   man/fetchnews.8 \
 400 #   man/nntpd.8
 401
 402 am__append_99 = \
 403     man/ctl_zoneinfo.8 \
 404     man/httpd.8
```

`am__append_96 = lib/map_nommap.c` is it somehow related to the fact that OpenBSD doesn't have a correct mmap implementation?

**Comment by elliefm** [comment]
> ```
> #define HAVE_DECLARE_OPTIMIZE /**/
> ```

Okay, so that means that our detection thinks your compiler does support this attribute.  So we need to figure out what's wrong with our detection.  Can you check your `config.log` for "checking for optimisation support" (n.b. en-uk spelling of "optimisation"!).  I get this:

```
configure:27250: checking for optimisation support
configure:27273: gcc -c -g -O0 -Wall -Wextra -Werror -Wl,--as-needed -fstack-protector-all    conftest.c >&5
configure:27273: $? = 0
configure:27281: result: yes
```

What do you get?

> Just in case the line numbers reported are somehow wrong, I'm pasting here some lines before and after.

Mmm, yeah, I don't see anything that looks like a "suffix rule definition" here, much less one with prerequisites.  This just looks like a bunch of variable definitions, and is basically the same thing I see in mine.  I guess the line number is way off...

Found a definition for the error here: https://www.gnu.org/software/make/manual/html_node/Error-Messages.html

> According to POSIX, a suffix rule cannot contain prerequisites. If a rule that could be a suffix rule has prerequisites it is interpreted as a simple explicit rule, with an odd target name. This requirement is obeyed when POSIX-conforming mode is enabled (the .POSIX target is defined). In versions of GNU make prior to 4.3, no warning was emitted and a suffix rule was created, however all prerequisites were ignored and were not part of the suffix rule. Starting with GNU make 4.3 the behavior is the same, and in addition this warning is generated. In a future version the POSIX-conforming behavior will be the only behavior: no rule with a prerequisite can be suffix rule and this warning will be removed.

"In versions of GNU make prior to 4.3, no warning was emitted" -> ahh.  My `make` is "GNU Make 4.1", so I'll never see this warning anyway.  The warning doesn't change the behaviour though, so I think it's safe to ignore for now -- but in the future the behaviour will change to match POSIX, which might break things when that happens...

I have a feeling this is an "automake is generating stuff that make complains about now" problem, and will probably be fixed (eventually) by upgrading to a version of automake that produces improved output... but I don't know if such a version already exists/etc.  For whatever it's worth, I have "automake (GNU automake) 1.15".

I had a skim through my generated Makefile, and none of the suffix rules have prerequisites, so??? 🤷‍♀️

**Comment by anatoli26** [comment]
@elliefm,

> Can you check your `config.log` for "checking for optimisation support"

```
6947:configure:26774: checking for optimisation support
6948-configure:26797: cc -c -O2 -pipe -g  -I/usr/local/include conftest.c >&5
6949-conftest.c:124:41: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
6950-    __attribute__((pure, always_inline, optimize("-O3")));
6951-                                        ^
6952-1 warning generated.
6953-configure:26797: $? = 0
6954-configure:26805: result: yes
```

In the Makefile generated on my OpenBSD, I see the following lines:
```
26774 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for optimisation support" >&5
26775 $as_echo_n "checking for optimisation support... " >&6; }
26776 if ${cyrus_cv_declare_optimize+:} false; then :
26777   $as_echo_n "(cached) " >&6
26778 else
26779   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
26780 /* end confdefs.h.  */
26781
26782 #include <stdlib.h>
26783 static inline int numcmp(int n1, int n2)
26784     __attribute__((pure, always_inline, optimize("-O3")));
26785 static int numcmp(int n1, int n2) { return n1 - n2; }
26786
26787 int
26788 main ()
26789 {
26790
26791     if (numcmp(1, 2)) exit(1);
26792
26793   ;
26794   return 0;
26795 }
26796 _ACEOF
26797 if ac_fn_c_try_compile "$LINENO"; then :
26798   cyrus_cv_declare_optimize=yes
26799 else
26800   cyrus_cv_declare_optimize=no
26801 fi
26802 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
26803 fi
26804
26805 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $cyrus_cv_declare_optimize" >&5
26806 $as_echo "$cyrus_cv_declare_optimize" >&6; }
26807 if test "$cyrus_cv_declare_optimize" = "yes"; then
26808
26809 $as_echo "#define HAVE_DECLARE_OPTIMIZE /**/" >>confdefs.h
26810
26811 fi
```
So what it tries to do is it tries to compile the code with the attribute, and the compilation actually succeeds, as a warning is not an error. Maybe it should be tried with a flag converting all warnings into errors?

> I had a skim through my generated Makefile, and none of the suffix rules have prerequisites, so??? :woman_shrugging:

hahaa ok

FWIW on OpenBSD we have autoconf 2.69, automake 1.16 and clang 10.0.1.

**Comment by elliefm** [comment]
> So what it tries to do is it tries to compile the code with the attribute, and the compilation actually succeeds, as a warning is not an error. Maybe it should be tried with a flag converting all warnings into errors?

Yep, that's exactly what I was thinking, just wanted to make sure that the optimize warning was the _only_ warning, so that I don't just introduce another annoying inconsistency to trip over later. :)

**Comment by anatoli26** [comment]
@elliefm, just in case you see something else strange, I'm pasting here the output of `grep -B3 -A10 warning config.log`:
```
1515-configure:6130: result: yes
1516-configure:6144: checking for library containing strerror
1517-configure:6175: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c  >&5
1518:conftest.c:34:6: warning: incompatible redeclaration of library function 'strerror' [-Wincompatible-library-redeclaration]
1519-char strerror ();
1520-     ^
1521-conftest.c:34:6: note: 'strerror' is a builtin with type 'char *(int)'
1522:1 warning generated.
1523-configure:6175: $? = 0
1524-configure:6192: result: none required
1525-configure:6204: checking for gawk
1526-configure:6231: result: awk
1527-configure:6242: checking for long file names
1528-configure:6274: result: yes
1529-configure:6285: checking for inline
1530-configure:6301: cc -c -O2 -pipe -g -I/usr/local/include conftest.c >&5
1531-configure:6301: $? = 0
1532-configure:6309: result: inline
--
1927-configure:17212: result: 4
1928-configure:17225: checking for runpath switch
1929-configure:17245: cc -o conftest -O2 -pipe -g -I/usr/local/include -R /usr/lib conftest.c  >&5
1930:warning: unknown remark option '-R' [-Wunknown-warning-option]
1931:1 warning generated.
1932-ld: error: cannot open /usr/lib: Is a directory
1933-cc: error: linker command failed with exit code 1 (use -v to see invocation)
1934-configure:17245: $? = 1
1935-configure: failed program was:
1936-| /* confdefs.h */
1937-| #define PACKAGE_NAME "cyrus-imapd"
1938-| #define PACKAGE_TARNAME "cyrus-imapd"
1939-| #define PACKAGE_VERSION "3.2.4"
1940-| #define PACKAGE_STRING "cyrus-imapd 3.2.4"
1941-| #define PACKAGE_BUGREPORT "https://github.com/cyrusimap/cyrus-imapd/issues"
--
5329-configure:20733: result: yes
5330-configure:20887: checking nonblocking method
5331-configure:20908: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5332:conftest.c:100:21: warning: expression result unused [-Wunused-value]
5333-fcntl(0, F_GETFL, 0)&FNDELAY
5334-~~~~~~~~~~~~~~~~~~~~^~~~~~~~
5335:1 warning generated.
5336-configure:20908: $? = 0
5337-configure:20925: result: fcntl
5338-configure:20928: checking timezone GMT offset method
5339-configure:20947: cc -c -O2 -pipe -g -I/usr/local/include conftest.c >&5
5340-configure:20947: $? = 0
5341-configure:20963: result: tm
5342-configure:20965: checking for shared mmap
5343-configure:21004: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5344-configure:21004: $? = 0
5345-configure:21004: ./conftest
--
5592-| }
5593-|
5594-configure:21071: result: no
5595:configure:21076: WARNING: *** This system does not have a working mmap()
5596:configure:21078: WARNING: *** Expect a considerable performance penalty
5597-configure:21115: checking for fcntl
5598-configure:21115: result: yes
5599-configure:21144: checking for library containing fdatasync
5600-configure:21175: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5601-configure:21175: $? = 0
5602-configure:21192: result: none required
5603-configure:21205: checking for library containing clock_gettime
5604-configure:21236: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5605-configure:21236: $? = 0
5606-configure:21253: result: none required
--
6946-configure:26766: result: no
6947-configure:26774: checking for optimisation support
6948-configure:26797: cc -c -O2 -pipe -g  -I/usr/local/include conftest.c >&5
6949:conftest.c:124:41: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
6950-    __attribute__((pure, always_inline, optimize("-O3")));
6951-                                        ^
6952:1 warning generated.
6953-configure:26797: $? = 0
6954-configure:26805: result: yes
6955:configure:26912: WARNING: No sphinx-build, won't be able to regenerate docs
6956-configure:26967: checking for perl module Pod::POM::View::Restructured
6957-configure:26973: result: no
6958:configure:26988: WARNING: No Pod::POM::View::Restructured, won't be able to regenerate docs
6959-configure:27009: checking python module: git
6960-configure:27018: result: no
6961:configure:27033: WARNING: GitPython not found, won't be able to regenerate docs
6962-configure:27190: checking that generated files are newer than configure
6963-configure:27196: result: done
6964-configure:27400: creating ./config.status
6965-
6966-## ---------------------- ##
6967-## Running config.status. ##
6968-## ---------------------- ##
6969-
6970-This file was extended by cyrus-imapd config.status 3.2.4, which was
6971-generated by GNU Autoconf 2.69.  Invocation command line was
--
6997-config.status:1741: executing perl/annotator/Makefile commands
6998-config.status:1741: executing perl/imap/Makefile commands
6999-config.status:1741: executing perl/sieve/managesieve/Makefile commands
7000:configure:30009: WARNING: unrecognized options: --disable-gtk-doc
7001-
7002-## ---------------- ##
7003-## Cache variables. ##
7004-## ---------------- ##
7005-
7006-ac_cv_accept_arg=socklen_t
7007-ac_cv_alignof_uint32_t=4
7008-ac_cv_argtype_accept3='socklen_t *'
7009-ac_cv_argtype_poll2=nfds_t
7010-ac_cv_atexit=atexit
```

**Comment by anatoli26** [comment]
@elliefm, just in case you see something else strange, I'm pasting here the output of `grep -B3 -A10 warning config.log`:
```
1515-configure:6130: result: yes
1516-configure:6144: checking for library containing strerror
1517-configure:6175: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c  >&5
1518:conftest.c:34:6: warning: incompatible redeclaration of library function 'strerror' [-Wincompatible-library-redeclaration]
1519-char strerror ();
1520-     ^
1521-conftest.c:34:6: note: 'strerror' is a builtin with type 'char *(int)'
1522:1 warning generated.
1523-configure:6175: $? = 0
1524-configure:6192: result: none required
1525-configure:6204: checking for gawk
1526-configure:6231: result: awk
1527-configure:6242: checking for long file names
1528-configure:6274: result: yes
1529-configure:6285: checking for inline
1530-configure:6301: cc -c -O2 -pipe -g -I/usr/local/include conftest.c >&5
1531-configure:6301: $? = 0
1532-configure:6309: result: inline
--
1927-configure:17212: result: 4
1928-configure:17225: checking for runpath switch
1929-configure:17245: cc -o conftest -O2 -pipe -g -I/usr/local/include -R /usr/lib conftest.c  >&5
1930:warning: unknown remark option '-R' [-Wunknown-warning-option]
1931:1 warning generated.
1932-ld: error: cannot open /usr/lib: Is a directory
1933-cc: error: linker command failed with exit code 1 (use -v to see invocation)
1934-configure:17245: $? = 1
1935-configure: failed program was:
1936-| /* confdefs.h */
1937-| #define PACKAGE_NAME "cyrus-imapd"
1938-| #define PACKAGE_TARNAME "cyrus-imapd"
1939-| #define PACKAGE_VERSION "3.2.4"
1940-| #define PACKAGE_STRING "cyrus-imapd 3.2.4"
1941-| #define PACKAGE_BUGREPORT "https://github.com/cyrusimap/cyrus-imapd/issues"
--
5329-configure:20733: result: yes
5330-configure:20887: checking nonblocking method
5331-configure:20908: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5332:conftest.c:100:21: warning: expression result unused [-Wunused-value]
5333-fcntl(0, F_GETFL, 0)&FNDELAY
5334-~~~~~~~~~~~~~~~~~~~~^~~~~~~~
5335:1 warning generated.
5336-configure:20908: $? = 0
5337-configure:20925: result: fcntl
5338-configure:20928: checking timezone GMT offset method
5339-configure:20947: cc -c -O2 -pipe -g -I/usr/local/include conftest.c >&5
5340-configure:20947: $? = 0
5341-configure:20963: result: tm
5342-configure:20965: checking for shared mmap
5343-configure:21004: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5344-configure:21004: $? = 0
5345-configure:21004: ./conftest
--
5592-| }
5593-|
5594-configure:21071: result: no
5595:configure:21076: WARNING: *** This system does not have a working mmap()
5596:configure:21078: WARNING: *** Expect a considerable performance penalty
5597-configure:21115: checking for fcntl
5598-configure:21115: result: yes
5599-configure:21144: checking for library containing fdatasync
5600-configure:21175: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5601-configure:21175: $? = 0
5602-configure:21192: result: none required
5603-configure:21205: checking for library containing clock_gettime
5604-configure:21236: cc -o conftest -O2 -pipe -g -I/usr/local/include -L/usr/local/lib conftest.c   -lpcre -lpcreposix >&5
5605-configure:21236: $? = 0
5606-configure:21253: result: none required
--
6946-configure:26766: result: no
6947-configure:26774: checking for optimisation support
6948-configure:26797: cc -c -O2 -pipe -g  -I/usr/local/include conftest.c >&5
6949:conftest.c:124:41: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
6950-    __attribute__((pure, always_inline, optimize("-O3")));
6951-                                        ^
6952:1 warning generated.
6953-configure:26797: $? = 0
6954-configure:26805: result: yes
6955:configure:26912: WARNING: No sphinx-build, won't be able to regenerate docs
6956-configure:26967: checking for perl module Pod::POM::View::Restructured
6957-configure:26973: result: no
6958:configure:26988: WARNING: No Pod::POM::View::Restructured, won't be able to regenerate docs
6959-configure:27009: checking python module: git
6960-configure:27018: result: no
6961:configure:27033: WARNING: GitPython not found, won't be able to regenerate docs
6962-configure:27190: checking that generated files are newer than configure
6963-configure:27196: result: done
6964-configure:27400: creating ./config.status
6965-
6966-## ---------------------- ##
6967-## Running config.status. ##
6968-## ---------------------- ##
6969-
6970-This file was extended by cyrus-imapd config.status 3.2.4, which was
6971-generated by GNU Autoconf 2.69.  Invocation command line was
--
6997-config.status:1741: executing perl/annotator/Makefile commands
6998-config.status:1741: executing perl/imap/Makefile commands
6999-config.status:1741: executing perl/sieve/managesieve/Makefile commands
7000:configure:30009: WARNING: unrecognized options: --disable-gtk-doc
7001-
7002-## ---------------- ##
7003-## Cache variables. ##
7004-## ---------------- ##
7005-
7006-ac_cv_accept_arg=socklen_t
7007-ac_cv_alignof_uint32_t=4
7008-ac_cv_argtype_accept3='socklen_t *'
7009-ac_cv_argtype_poll2=nfds_t
7010-ac_cv_atexit=atexit
```

**Comment by elliefm** [comment]
I think this should fix the detection of the missing optimisation support, and so fix the "unknown attribute" warnings:

https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v33/3265-detect-optimize-support

and as a .patch:

https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v33/3265-detect-optimize-support.patch

Can you try it out and report back?  I'd like to see the output from config.log again -- I'm expecting it to be about the same, except with an error instead of a warning, and "result: no"

**Comment by anatoli26** [comment]
Working!

`./configure` output is: `909:checking for optimisation support... no`

Here it goes:
```
6947:configure:26774: checking for optimisation support
6948-configure:26805: cc -c -O2 -pipe -g -Werror  -I/usr/local/include conftest.c >&5
6949-conftest.c:124:61: error: unknown attribute 'optimize' ignored [-Werror,-Wunknown-attributes]
6950-                        __attribute__((pure, always_inline, optimize("-O3")));
6951-                                                            ^
6952-1 error generated.
6953-configure:26805: $? = 1
6954-configure: failed program was:
6955-| /* confdefs.h */
6956-| #define PACKAGE_NAME "cyrus-imapd"
6957-| #define PACKAGE_TARNAME "cyrus-imapd"
6958-| #define PACKAGE_VERSION "3.2.4"
6959-| #define PACKAGE_STRING "cyrus-imapd 3.2.4"
6960-| #define PACKAGE_BUGREPORT "https://github.com/cyrusimap/cyrus-imapd/issues"
6961-| #define PACKAGE_URL "https://www.cyrusimap.org"
6962-| #define PACKAGE "cyrus-imapd"
6963-| #define VERSION "3.2.4"
6964-| #define CYRUS_USER "_cyrus"
6965-| #define STDC_HEADERS 1
6966-| #define HAVE_SYS_TYPES_H 1
6967-| #define HAVE_SYS_STAT_H 1
6968-| #define HAVE_STDLIB_H 1
6969-| #define HAVE_STRING_H 1
6970-| #define HAVE_MEMORY_H 1
6971-| #define HAVE_STRINGS_H 1
6972-| #define HAVE_INTTYPES_H 1
6973-| #define HAVE_STDINT_H 1
6974-| #define HAVE_UNISTD_H 1
6975-| #define __EXTENSIONS__ 1
6976-| #define _ALL_SOURCE 1
6977-| #define _GNU_SOURCE 1
6978-| #define _POSIX_PTHREAD_SEMANTICS 1
6979-| #define _TANDEM_SOURCE 1
6980-| #define HAVE_LONG_FILE_NAMES 1
6981-| #define HAVE_VISIBILITY 1
6982-| #define HAVE_DLFCN_H 1
6983-| #define LT_OBJDIR ".libs/"
6984-| #define SIZEOF_INT 4
6985-| #define SIZEOF_LONG 8
6986-| #define SIZEOF_SIZE_T 8
6987-| #define SIZEOF_OFF_T 8
6988-| #define SIZEOF_TIME_T 8
6989-| #define SIZEOF_LONG_LONG_INT 8
6990-| #define SIZEOF_UNSIGNED_LONG_LONG_INT 8
6991-| #define ALIGNOF_UINT32_T 4
6992-| #define HAVE_UNISTD_H 1
6993-| #define HAVE_SYS_SELECT_H 1
6994-| #define HAVE_SYS_PARAM_H 1
6995-| #define HAVE_STDARG_H 1
6996-| #define HAVE_MEMMOVE 1
6997-| #define HAVE_STRCASECMP 1
6998-| #define HAVE_FTRUNCATE 1
6999-| #define HAVE_STRERROR 1
7000-| #define HAVE_STRSEP 1
7001-| #define HAVE_MEMMEM 1
7002-| #define HAVE_STRLCAT 1
7003-| #define HAVE_STRLCPY 1
7004-| #define HAVE_GETGROUPLIST 1
7005-| #define HAVE_FMEMOPEN 1
7006-| #define HAVE_PSELECT 1
7007-| #define HAVE_FUTIMENS 1
7008-| #define HAVE_FUTIMES 1
7009-| #define HAVE_DIRENT_H 1
7010-| #define HAVE_STDLIB_H 1
7011-| #define HAVE_DN_EXPAND 1
7012-| #define HAVE_GETADDRINFO /**/
7013-| #define HAVE_GETNAMEINFO /**/
7014-| #define HAVE_SS_FAMILY /**/
7015-| #define HAVE_SOCKADDR_SA_LEN /**/
7016-| #define HAVE_SYS_TIME_H 1
7017-| #define TIME_WITH_SYS_TIME 1
7018-| #define HAVE_STRUCT_TM_TM_ZONE 1
7019-| #define HAVE_TM_ZONE 1
7020-| #define HAVE_TIMEGM 1
7021-| #define HAVE_VPRINTF 1
7022-| #define TIME_T_FMT "%lld"
7023-| #define strtotimet(a,b,c) strtoll(a,b,c)
7024-| #define SIZE_T_FMT "%lu"
7025-| #define strtosizet(a,b,c) strtoul(a,b,c)
7026-| #define OFF_T_FMT "%lld"
7027-| #define strtoofft(a,b,c) strtoll(a,b,c)
7028-| #define XAPIAN_CJK_TOKENS "none"
7029-| #define USE_SIEVE /**/
7030-| #define YYTEXT_POINTER 1
7031-| #define ENABLE_REGEX /**/
7032-| #define HAVE_PCREPOSIX_H /**/
7033-| #define HAS_STRERROR /**/
7034-| #define HAVE_SYS_RESOURCE_H 1
7035-| #define HAVE_SETRLIMIT 1
7036-| #define HAVE_GETRLIMIT 1
7037-| #define HAVE_DAEMON 1
7038-| #define HAVE_SETSID 1
7039-| #define HAVE_SHUTDOWN 1
7040-| #define HAVE_SOCKLEN_T /**/
7041-| #define HAVE_STRUCT_SOCKADDR_STORAGE /**/
7042-| #define HAVE_RLIM_T /**/
7043-| #define SPT_TYPE SPT_BUILTIN
7044-| #define HAVE_FDATASYNC /**/
7045-| #define HAVE_LIBUUID /**/
7046-| #define HAVE_SSL /**/
7047-| #define HAVE_TLS_ALPN /**/
7048-| #define HAVE_ZLIB /**/
7049-| #define HAVE_JANSSON /**/
7050-| #define HAVE_LIBJANSSON 1
7051-| #define HAVE_ICU /**/
7052-| #define MASTER_PIDFILE "/var/run/cyrus-master.pid"
7053-| #define USE_AUTOCREATE /**/
7054-| #define USE_IDLED /**/
7055-| #define WITH_DAV /**/
7056-| #define HAVE_XML2 /**/
7057-| #define HAVE_XML_FIRSTCHILD /**/
7058-| #define HAVE_XML_BUFFERDETACH /**/
7059-| #define HAVE_ICAL /**/
7060-| #define HAVE_RFC7986_COLOR /**/
7061-| #define HAVE_RSCALE /**/
7062-| #define HAVE_NGHTTP2 /**/
7063-| #define HAVE_DECL_NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL 1
7064-| #define HAVE_WSLAY /**/
7065-| #define HAVE_BROTLI /**/
7066-| #define HAVE_ZSTD /**/
7067-| #define USE_HTTPD /**/
7068-| #define HAVE_ET_COM_ERR_H /**/
7069-| #define SYSLOG_FACILITY LOG_MAIL
7070-| #define HAVE_GETDTABLESIZE 1
7071-| #define HAVE_APOP /**/
7072-| #define SIZEOF_LONG 8
7073-| #define HAVE_SQLITE /**/
7074-| /* end confdefs.h.  */
7075-|
7076-|                     #include <stdlib.h>
7077-|                     static inline int numcmp(int n1, int n2)
7078-|                         __attribute__((pure, always_inline, optimize("-O3")));
7079-|                     static int numcmp(int n1, int n2) { return n1 - n2; }
7080-|
7081-| int
7082-| main ()
7083-| {
7084-|  if (numcmp(1, 2)) exit(1);
7085-|   ;
7086-|   return 0;
7087-| }
7088-|
7089-configure:26820: result: no
```

**Comment by anatoli26** [comment]
@elliefm, could you please cherry-pick this to the 3.2 branch?

**Comment by elliefm** [comment]
Fixed on 3.2 and master branches by #3276
